src/h3/frame.rs

```diff
@@ -0,0 +1,856 @@
+        }
+
+        Ok(before - b.cap())
+    }
+}
+
+impl std::fmt::Debug for H3Frame {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            H3Frame::Data { payload } => {
+                write!(f, "DATA len={}", payload.len())?;
+            },
+
+            H3Frame::Headers { header_block } => {
+                write!(f, "HEADERS len={}", header_block.len())?;
+            },
+
+            H3Frame::Priority {
+                priority_elem,
+                elem_dependency,
+                prioritized_element_id,
+                element_dependency_id,
+                weight,
+            } => {
+                let peid = match prioritized_element_id {
+                    Some(x) => x.to_string(),
+                    None => String::from("None"),
+                };
+
+                let edid = match element_dependency_id {
+                    Some(x) => x.to_string(),
+                    None => String::from("None"),
+                };
+
+                write!(f, "PRIORITY priority element type={:?} element dependency type={:?} prioritized element id={:?} element dependency id={:?} weight={:?}", priority_elem, elem_dependency, peid, edid, weight)?;
+            },
+
+            H3Frame::CancelPush { push_id } => {
+                write!(f, "CANCEL_PUSH push id={}", push_id)?;
+            },
+
+            H3Frame::Settings {
+                num_placeholders,
+                max_header_list_size,
+                qpack_max_table_capacity,
+                qpack_blocked_streams,
+            } => {
+                write!(f, "SETTINGS num placeholders={}, max header list size={}, qpack max table capacity={}, qpack blocked streams={} ", num_placeholders.unwrap_or(999), max_header_list_size.unwrap_or(999), qpack_max_table_capacity.unwrap_or(999), qpack_blocked_streams.unwrap_or(999) )?;
+            },
+
+            H3Frame::PushPromise {
+                push_id,
+                header_block,
+            } => {
+                write!(
+                    f,
+                    "PUSH_PROMISE push id={} len={}",
+                    push_id,
+                    header_block.len()
+                )?;
+            },
+
+            H3Frame::GoAway { stream_id } => {
+                write!(f, "GOAWAY stream id={}", stream_id)?;
+            },
+
+            H3Frame::MaxPushId { push_id } => {
+                write!(f, "MAX_PUSH_ID push id={}", push_id)?;
+            },
+
+            H3Frame::DuplicatePush { push_id } => {
+                write!(f, "DUPLICATE_PUSH push id={}", push_id)?;
+            },
+        }
+
+        Ok(())
+    }
+}
+
+fn parse_settings_frame(
+    payload_length: u64, b: &mut octets::Octets,
+) -> Result<H3Frame> {
+    let mut num_placeholders = None;
+    let mut max_header_list_size = None;
+    let mut qpack_max_table_capacity = None;
+    let mut qpack_blocked_streams = None;
+
+    while b.off() < payload_length as usize {
```

Comment:
This only really works if the SETTINGS frame is at the start of the buffer. Though it would be fixed for free if you do the `b.get_bytes()` trick I suggested above.

Reply:
This can just be `while b.cap() > 0 {` now.